### ISSUE TITLE
Sidekiq severely slowing down when moving from single to multiple queues to process the same jobs

### ISSUE THREAD

**Issue by ddujisin** [mperham/sidekiq#4597]
We are running a web application on Ruby-on-rails on a load-balanced dual server setup. We are using sidekiq to run many small jobs in the background. Initially we only had one queue but we decided to add more queues to help process these jobs in parallel. Nothing in the nature of the jobs changed except they are running in parallel on multiple queues. We noticed a huge slowdown in processing each individual job, even though the servers are not even close to being overwhelmed. We’ve been playing with the pool values, priorities, number of queues, etc and have it working better but nowhere near the speed we need, some individual jobs are taking 40, 45 minutes but they should take less than a minute (and they do when running a single queue).

Ruby version: ruby 2.2.2p95
Sidekiq / Pro / Enterprise version(s): Sidekiq v4.2.10

Setup: AWS load balancer + 2 app servers (r4.4xlarge), Mysql db server. servers are running at about 10-15% CPU consistently with no peaks. DB server a bit busier, but still between 30-70% CPU on average, memory usage is at about 50% -60% consistently. We see roughly the same resource usage wether we are running a single or multiple queues.

**Sidekiq.yml**
---
:logfile: log/sidekiq.log
:concurrency: 5
production:
  :concurrency: 9
:queues:
  - ["event_process_queue", 2]
  - ["event_process_queue_0", 2]
  - ["event_process_queue_1", 2]
  - ["event_process_queue_2", 2]
  - ["event_process_queue_3", 2]
  - ["event_process_queue_4", 2]
  - ["default", 1]
:limits:
  default: 3
  event_process_queue: 1
  event_process_queue_0: 1
  event_process_queue_1: 1
  event_process_queue_2: 1
  event_process_queue_3: 1
  event_process_queue_4: 1

**database.yml**
production:
  <<: *default
  database: production
  pool: 25

**Comment by mperham** [comment]
Show me your sidekiq.rb initializer.

**Comment by mperham** [comment]
Can you share the source for a slow job that should be fast? I see you are using limits, that's not part of Sidekiq's supported functionality and is possibly a cause of the slowdown.

**Comment by ddujisin** [comment]
we don’t have a sidekiq.rb file in the initializer folder (only the ones in the gem folders)

Daniel Dujisin
SYNCrew




On Jun 15, 2020, at 13:08, Mike Perham <notifications@github.com> wrote:


Show me your sidekiq.rb initializer.

**Comment by ddujisin** [comment]
The jobs slowing down can happen in any of the queues. But the ones coming to the event_process queues are the ones we are concerned about. These are the ones that individually take very little time when running on a single queue and get delayed on multiple. queues. 

So you are saying we could try with no limits? we didn’t really want to have more than one job processing from the same queue, so I guess if we limit the concurrency to 6 that should give us one concurrent job per queue and we wouldn't have to use limits, which are as you said, unsupported?

Thanks, really appreciate your help!

Daniel Dujisin
SYNCrew




On Jun 15, 2020, at 13:09, Mike Perham <notifications@github.com> wrote:


Can you share the source for a slow job that should be fast? I see you are using limits, that's not part of Sidekiq's supported functionality and is possibly a cause of the slowdown.

**Comment by mperham** [comment]
I don't know what else could cause your slowdowns except the limiter you've imposed. It's always possible that you have a database or networking issue but that's impossible for me to tell. Usually people configure something wrongly and cause the massive slowdown.

Concurrency of six means you can have up to six jobs processing concurrently, i.e. at the same time. Sidekiq does not provide a way to limit concurrency per queue. Why are you trying to limit concurrency at all? Why not just throw everything into `default` and let Sidekiq process it as fast as possible?

**Comment by ddujisin** [comment]
The more jobs we have happening at the same time, the more it exponentially slows down. So for example, if it takes 100 jobs let’s say 10 minutes to complete when running one by one, if we use 5 queues with the same jobs being distributed through the 5 queues it will take 20 or more minutes to complete. And we do have jobs that are much higher priority, so we would want to take care of those before anything else. I ruled out networking issues since it’s all virtualized in Amazon Web services and we see no delays whatsoever when running single jobs.

What do you think about having 2 queues, one high priority and one low priority (default) and just have the concurrency set at 10 to start and see if we can increase?

Daniel Dujisin
SYNCrew




On Jun 15, 2020, at 13:56, Mike Perham <notifications@github.com> wrote:


I don't know what else could cause your slowdowns except the limiter you've imposed. It's always possible that you have a database or networking issue but that's impossible for me to tell. Usually people configure something wrongly and cause the massive slowdown.

Concurrency of six means you can have up to six jobs processing concurrently, i.e. at the same time. Sidekiq does not provide a way to limit concurrency per queue. Why are you trying to limit concurrency at all? Why not just throw everything into default and let Sidekiq process it as fast as possible?

**Comment by mperham** [comment]
Pardon the questions, I don't know your level of expertise. It sounds like you have bigger performance problems than Sidekiq. Are you missing a database index? Have you profiled and tuned your app code and know that's not the problem?

You are running Sidekiq on an instance with 16 cores but do you understand that Sidekiq will only use one core per process; you need to start up multiple Sidekiq processes if you want to process jobs on many cores?

I would start over: remove all concurrency, limits and queue config from the yml.  Stick with just the default queue and concurrency of 10, Sidekiq's defaults. Now if after that Sidekiq is slow, you need to run a profiler or add some debug logging to your app code and determine what is slow. I don't see any reason why Sidekiq is the cause here.

Something like [query_track](https://github.com/kirillshevch/query_track) can help.

**Comment by ddujisin** [comment]
Thanks for your help, sound like we have some stuff to try here. It’s been all a bit confusing because the only thing that changes when running multiple jobs is the time it takes to complete, with NO effect on server resource usage, that’s why we assumed the problem was with the sidekiq config and started playing with all the options in the yml files. 

Appreciate your help! Take care,  Daniel


On Jun 15, 2020, at 14:53, Mike Perham <notifications@github.com> wrote:


Pardon the questions, I don't know your level of expertise. It sounds like you have bigger performance problems than Sidekiq. Are you missing a database index? Have you profiled and tuned your app code and know that's not the problem?

You are running Sidekiq on an instance with 16 cores but do you understand that Sidekiq will only use one core per process; you need to start up multiple Sidekiq processes if you want to process jobs on many cores?

I would start over: remove all concurrency, limits and queue config from the yml. Stick with just the default queue and concurrency of 10, Sidekiq's defaults. Now if after that Sidekiq is slow, you need to run a profiler or add some debug logging to your app code and determine what is slow. I don't see any reason why Sidekiq is the cause here.

Something like query_track <https://github.com/kirillshevch/query_track> can help.

**Comment by ddujisin** [comment]
Hi,

We made the change to run 10 concurrent jobs in the default queue (with no other queues) and it got significantly worse compared to the previous setup we had with multiple queues and limits. We did not change anything else in the code, just the queue setup. 

I suspect that most people that use this setup have jobs that are very similar to each other in terms of load. This is not our case, some jobs will take longer, and that's because they are running more calculations - no mystery there, we know this.

So I have a suspicion - not based on evidence, just a hunch - that having jobs that require much more computing power and time running in parallel with jobs which require much less of that makes those quicker jobs slower and that’s why everything slows down. I’ve seen the server process many jobs very quickly when they are ALL small. Has anybody observed something like this?


On Jun 15, 2020, at 15:08, Daniel Dujisin <daniel@syncrew.com> wrote:

Thanks for your help, sound like we have some stuff to try here. It’s been all a bit confusing because the only thing that changes when running multiple jobs is the time it takes to complete, with NO effect on server resource usage, that’s why we assumed the problem was with the sidekiq config and started playing with all the options in the yml files. 

Appreciate your help! Take care,  Daniel


On Jun 15, 2020, at 14:53, Mike Perham <notifications@github.com <mailto:notifications@github.com>> wrote:


Pardon the questions, I don't know your level of expertise. It sounds like you have bigger performance problems than Sidekiq. Are you missing a database index? Have you profiled and tuned your app code and know that's not the problem?

You are running Sidekiq on an instance with 16 cores but do you understand that Sidekiq will only use one core per process; you need to start up multiple Sidekiq processes if you want to process jobs on many cores?

I would start over: remove all concurrency, limits and queue config from the yml. Stick with just the default queue and concurrency of 10, Sidekiq's defaults. Now if after that Sidekiq is slow, you need to run a profiler or add some debug logging to your app code and determine what is slow. I don't see any reason why Sidekiq is the cause here.

Something like query_track <https://github.com/kirillshevch/query_track> can help.

**Comment by mperham** [comment]
If you have compute heavy jobs, you need to lower concurrency and increase processes. Sidekiq Enterprise’s multi process feature can help here or you can roll your own.
